src/training/final.ts

```diff
@@ -27,13 +27,208 @@ type Coupon = {
   amount: number;
 };
 
+// enumではなくこちらを使うコーディングは要調査
+const PAYMENT_METHOD= {
+   現金:{
+      code: 0
+   },
+   クレジットカード: {
+      code: 1
+   },
+   銀行振り込み: {
+      code: 2
+   }
+} as const;
+
+enum PAYMENT_METHOD_ENUM {
+   現金=0,
+   クレジットカード=1,
+   銀行振り込み=2,
+}
+
+enum TRANSACTION_TYPE_ENUM {
+   決済 = 0,
+   チャージ = 1,
+}
+
+const coupon_100: Coupon = { couponId: 1, couponName: "coupon_100",amount: 100 };
+const coupon_500: Coupon = { couponId: 2, couponName: "coupon_500", amount: 500 };
+const coupon_1000: Coupon = { couponId: 3, couponName: "coupon_1000", amount: 1000 };
+
 type Wallet = {
-  walletId;
-  username;
-  balance;
-  method;
-  transactionHistroy;
-  coupon;
+  walletId: number;
+  username: string;
+  balance: number;
+  method: 0 | 1 | 2;
+  transactionHistory: Transaction[];
+  coupon?: Coupon[];
 };
 
-type Transaction = {};
+type Transaction = {
+   walletId: number;
+   paymentAmount: number;
+   transactionType: TRANSACTION_TYPE_ENUM.決済 | TRANSACTION_TYPE_ENUM.チャージ;
+   method: PAYMENT_METHOD_ENUM.現金 | PAYMENT_METHOD_ENUM.クレジットカード | PAYMENT_METHOD_ENUM.銀行振り込み;
+   transactionDate: Date;
+   balance: number;
+};
+
+export const userWallets: Wallet[] = [
+   {walletId:1, username: "田中一郎", balance: 10000, method: 1, transactionHistory: []},
+   {walletId:2, username: "田中二郎", balance: 10000, method: 0, transactionHistory: [], coupon: [coupon_100]},
+   {walletId:3, username: "田中三郎", balance: 10000, method: 2, transactionHistory: [], coupon: [coupon_100, coupon_500]},
+   {walletId:4, username: "田中四郎", balance: 10000, method: 1, transactionHistory: [], coupon: [coupon_100, coupon_1000]},
+   {walletId:5, username: "田中五郎", balance: 10000, method: 0, transactionHistory: [], coupon: [coupon_100, coupon_500,coupon_1000]}
+];
+
+export const doSettlement = (walletId: number, paymentAmount: number): void => {
+
+   // 処理対象のウォレットのインデックスを所得する。
+   const getIndex = (): number => { 
+      const indices = userWallets.map( (userWallet,index) => {
+         if (userWallet.walletId === walletId && userWallet.balance >= paymentAmount) {
+            return index;
+         }
+         return -999;
+      }).filter(index => index !== -999);
+      return indices[0];
+   }
+   // 割引額を取得する。
+   const getCouponAmount = (targetIndex: number): number => {
+
+      const getTargetCouponAmount = ():number => {
+         if (userWallets[targetIndex].coupon) {
+            const couponAmount = userWallets[targetIndex].coupon.reduce(
+               (max, coupon) => Math.max(max, coupon.amount as unknown as number), 0);
+
+            const unuseCouponList: Coupon[] = [];
+            userWallets[targetIndex].coupon.forEach( (couponItem) => {
+               if (couponItem.amount !== couponAmount) unuseCouponList.push(couponItem);
+            });
```

Comment:
これfilterの方がきれいに書けると思いますがどうでしょうか
```suggestion
             const unuseCouponList = userWallets[targetIndex].coupon.filter( (item) => item.amount !== couponAmount);
```

Reply:
提示していただいたコードの方が処理がわかりやすいため、コードを変更ました。